File: src/main/java/dark/leech/text/plugin/js/loader/AbstractLoader.java
package dark.leech.text.plugin.js.loader;

import dark.leech.text.action.Log;
import dark.leech.text.enities.PluginEntity;
import dark.leech.text.plugin.js.sandbox.JsSandbox;

/**
 * Abstract base loader for vBook plugins using sandbox pattern. Extracts common sandbox creation,
 * script execution, and error handling patterns.
 */
public abstract class AbstractLoader<T> {

    protected final PluginEntity plugin;

    protected AbstractLoader(PluginEntity plugin) {
        this.plugin = plugin;
    }

    /**
     * Get the script content to execute.
     *
     * @return Script string or null if not available
     */
    protected abstract String getScript();

    /**
     * Get the loader type for sandbox isolation.
     *
     * @return Loader type enum (LIST, DETAIL, TEXT)
     */
    protected abstract LoaderType getLoaderType();

    /**
     * Get the script name for logging.
     *
     * @return Script name (e.g., "tocGetter", "detailGetter")
     */
    protected String getScriptName() {
        return getLoaderType().getScriptName();
    }

    /**
     * Process the result from execute() function call.
     *
     * @param result The result object from JavaScript execution
     * @param url The target URL
     * @return Processed result or null on error
     */
    protected abstract T processResult(Object result, String url);

    /**
     * Load data from URL using the configured script in a secure sandbox.
     *
     * @param url Target URL
     * @return Processed result or null on error
     */
    public T load(String url) {
        String script = getScript();
        if (plugin == null || script == null) {
            return null;
        }

        JsSandbox sandbox = null;
        try {
            // Extract base URL
            String baseUrl = plugin.getSource();
            if (baseUrl == null || baseUrl.isEmpty()) {
                baseUrl = JsApiSetup.extractBaseUrl(url);
            }

            // Create secure sandbox for this loader type
            JsSandbox.Builder builder =
                    new JsSandbox.Builder()
                            .loaderType(getLoaderType())
                            .baseUrl(baseUrl)
                            .targetUrl(url);

            // Add plugin source if available
            if (plugin.getSource() != null && !plugin.getSource().isEmpty()) {
                builder.pluginSource(plugin.getSource());
            }

            sandbox = builder.build();

            // Execute plugin script in sandbox
            if (!sandbox.execute(script, getScriptName())) {
                return null;
            }

            // Call execute(url) function in sandbox
            Object result = sandbox.callFunction("execute", url);

            if (result == null || result == org.mozilla.javascript.Undefined.instance) {
                return null;
            }

            return processResult(result, url);

        } catch (Exception e) {
            String errorMsg = e.getMessage();
            if (errorMsg == null) {
                errorMsg = e.getClass().getName();
            }
            Log.add("JavaScript execution error in " + getScriptName() + ": " + errorMsg);
            return null;
        } finally {
            if (sandbox != null) {
                sandbox.close();
            }
        }
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/js/loader/AbstractLoader.kt
package dev.haipham22.leechtext.plugin.js.loader

import dev.haipham22.leechtext.EngineDispatchers
import dev.haipham22.leechtext.entities.PluginEntity
import dev.haipham22.leechtext.log.EngineLogger
import dev.haipham22.leechtext.plugin.js.sandbox.JsSandbox
import dev.haipham22.leechtext.plugin.js.sandbox.createSandbox
import dev.haipham22.leechtext.util.parseBaseUrl

/**
 * Base loader cho vBook plugin (port từ loader/AbstractLoader.java; P5.2c common —
 * engine qua JsSandbox expect/actual, Rhino context thread-bound nên load() vẫn chạy
 * trong EngineDispatchers.withEngine).
 */
abstract class AbstractLoader<T>(
    protected val log: EngineLogger,
    protected val plugin: PluginEntity,
) {
    /** Script content sẽ execute, null nếu plugin không có. */
    protected abstract fun getScript(): String?

    /** Loại loader cho sandbox isolation. */
    protected abstract fun getLoaderType(): LoaderType

    /** Tên script cho log (vd "tocGetter", "detailGetter"). */
    protected open fun getScriptName(): String = getLoaderType().scriptName

    /** Xử lý kết quả từ call execute() function. */
    protected abstract fun processResult(
        result: Any?,
        url: String?,
    ): T?

    /** Load data từ URL bằng script của plugin trong sandbox bảo mật. Null trên error. */
    open suspend fun load(url: String?): T? = EngineDispatchers.withEngine { loadInSandbox(url) }

    /** Thân legacy AbstractLoader.load — chạy trên engine dispatcher. */
    @Suppress("ReturnCount") // guard-chain: từng bước fail thoát sớm, lifecycle sandbox/builder giữ nguyên shape
    protected open fun loadInSandbox(url: String?): T? {
        val script = getScript() ?: return null

        var sandbox: JsSandbox? = null
        try {
            // Extract base URL
            val baseUrl =
                plugin.source?.takeIf { it.isNotEmpty() } ?: parseBaseUrl(url ?: "")

            // Create secure sandbox cho loader type này
            sandbox =
                createSandbox(log) {
                    loaderType = getLoaderType()
                    this.baseUrl = baseUrl
                    targetUrl = url
                    if (!plugin.source.isNullOrEmpty()) pluginSource = plugin.source
                    // Plugin config (vBook extension-api) — inject làm JS constants
                    if (!plugin.config.isNullOrEmpty()) pluginConfig = plugin.config
                    if (!plugin.extraScripts.isNullOrEmpty()) extraScripts = plugin.extraScripts
                }

            // Execute plugin script trong sandbox
            if (!sandbox.execute(script, getScriptName())) return null

            // Call execute(url) trong sandbox — kết quả đã normalize
            val result = sandbox.callFunction("execute", url)
            if (result == null) return null

            return processResult(result, url)
        } catch (e: Exception) {
            val errorMsg = e.message ?: e::class.simpleName
            log.add("JavaScript execution error in ${getScriptName()}: $errorMsg")
            return null
        } finally {
            sandbox?.close()
        }
    }
}
